src/main/java/com/github/nkzawa/engineio/client/transports/WebSocket.java

```diff
@@ -37,70 +46,87 @@ protected void doOpen() {
-                    EventThread.exec(new Runnable() {
-                        @Override
-                        public void run() {
-                            Map<String, String> headers = new TreeMap<String, String>(String.CASE_INSENSITIVE_ORDER);
-                            Iterator<String> it = serverHandshake.iterateHttpFields();
-                            while (it.hasNext()) {
-                                String field = it.next();
-                                if (field == null) continue;
-                                headers.put(field, serverHandshake.getFieldValue(field));
-                            }
-                            self.emit(EVENT_RESPONSE_HEADERS, headers);
-
-                            self.onOpen();
-                        }
-                    });
-                }
-                @Override
-                public void onClose(int i, String s, boolean b) {
-                    EventThread.exec(new Runnable() {
-                        @Override
-                        public void run() {
-                            self.onClose();
-                        }
-                    });
+        final OkHttpClient client = new OkHttpClient();
+        if (this.sslContext != null) {
+            SSLSocketFactory factory = sslContext.getSocketFactory();// (SSLSocketFactory) SSLSocketFactory.getDefault();
+            client.setSslSocketFactory(factory);
+        }
+        final Request request = new Request.Builder()
+                .url(uri())
+                .build();
+        WebSocketCall.create(client, request).enqueue(new WebSocketListener() {
+            @Override
+            public void onOpen(com.squareup.okhttp.ws.WebSocket webSocket, Request request, Response response) throws IOException {
+                ws = webSocket;
+                final Map<String, String> headers = new TreeMap<String, String>(String.CASE_INSENSITIVE_ORDER);
+                for (int i = 0, size = response.headers().size(); i < size; i++) {
+                    headers.put(response.headers().name(i), response.headers().value(i));
                 }
-                @Override
-                public void onMessage(final String s) {
-                    EventThread.exec(new Runnable() {
-                        @Override
-                        public void run() {
-                            self.onData(s);
-                        }
-                    });
+                EventThread.exec(new Runnable() {
+                    @Override
+                    public void run() {
+                        self.emit(EVENT_RESPONSE_HEADERS, headers);
+                        self.onOpen();
+                    }
+                });
+            }
+
+            @Override
+            public void onMessage(BufferedSource payload, PayloadType type) throws IOException {
+                Object data;
+                switch (type) {
+                    case TEXT:
+                        data = payload.readUtf8();
```

Comment:
We have to cache data before executing on EventThread. OkHttp have to call payload.close() before onMessage return.